dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/changelog/ChangeLogServiceTest.java

```diff
@@ -0,0 +1,311 @@
+        });
+  }
+
+  @Test
+  void shouldReturnChangeLogsWhenDataValueIsUpdatedTwiceInARow()
+      throws NotFoundException, IOException {
+    injectSecurityContextUser(manager.get(User.class, "M5zQapPyTZI"));
+
+    Event event = manager.get(Event.class, "QRYjLTiJTrA");
+    assertNotNull(event);
+    String dataElementUid = event.getEventDataValues().iterator().next().getDataElement();
+    DataElement dataElement = manager.get(DataElement.class, dataElementUid);
+    assertNotNull(dataElement);
+
+    TrackerObjects trackerObjects = fromJson("tracker/event_and_enrollment.json");
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "20");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "25");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+
+    List<EventChangeLog> changeLogs = changeLogService.getEventChangeLog(UID.of("QRYjLTiJTrA"));
+
+    assertAll(
+        () -> {
+          validateNonValueElements(3, changeLogs, importUser, dataElementUid);
+          assertEquals("25", changeLogs.get(0).change().dataValue().currentValue());
+          assertEquals("20", changeLogs.get(0).change().dataValue().previousValue());
+          assertEquals("20", changeLogs.get(1).change().dataValue().currentValue());
+          assertEquals("15", changeLogs.get(1).change().dataValue().previousValue());
+          assertEquals("15", changeLogs.get(2).change().dataValue().currentValue());
+          assertNull(changeLogs.get(2).change().dataValue().previousValue());
+        });
+  }
+
+  @Test
+  void shouldReturnChangeLogsWhenDataValueIsCreatedUpdatedAndDeleted()
+      throws IOException, NotFoundException {
+    injectSecurityContextUser(manager.get(User.class, "M5zQapPyTZI"));
+
+    Event event = manager.get(Event.class, "QRYjLTiJTrA");
+    assertNotNull(event);
+    String dataElementUid = event.getEventDataValues().iterator().next().getDataElement();
+    DataElement dataElement = manager.get(DataElement.class, dataElementUid);
+    assertNotNull(dataElement);
+
+    TrackerObjects trackerObjects = fromJson("tracker/event_and_enrollment.json");
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "20");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+
+    updateDataValue(trackerObjects, event.getUid(), dataElementUid, "");
+    assertNoErrors(trackerImportService.importTracker(importParams, trackerObjects));
+
+    List<EventChangeLog> changeLogs = changeLogService.getEventChangeLog(UID.of("QRYjLTiJTrA"));
+
+    assertAll(
+        () -> {
+          validateNonValueElements(3, changeLogs, importUser, dataElementUid);
+          assertNull(changeLogs.get(0).change().dataValue().currentValue());
+          assertEquals("20", changeLogs.get(0).change().dataValue().previousValue());
```

Comment:
I think an assertion like

`assertChange(change, prev, cur)`

would make this even better to read